collator/src/validator/impls/std_impl/session.rs

```diff
@@ -184,41 +183,49 @@ impl ValidatorSession {
 
         let state = &self.inner.state;
 
-        // Remove cached slot
+        // first check for prevent slow operations
+        if state.block_signatures.contains_key(&block_id.seqno) {
+            anyhow::bail!(
+                "block validation is already in progress. \
+                 session_id={}, block_id={:?}",
+                self.inner.session_id,
+                block_id
+            );
+        }
+
         let cached = state
             .cached_signatures
-            .peek(&block_id.seqno, &scc::ebr::Guard::new())
-            .map(Arc::clone);
+            .get(&block_id.seqno)
+            .map(|entry| entry.clone());
 
-        // Prepare block signatures
-        let block_signatures = match &cached {
-            Some(cached) => self.reuse_signatures(block_id, cached.clone()).await,
-            None => self.prepare_new_signatures(block_id),
-        }
-        .build(block_id, state.weight_threshold);
+        let block_signatures = {
+            match &cached {
+                Some(cached) => self.reuse_signatures(block_id, cached.clone()).await,
+                None => self.prepare_new_signatures(block_id),
+            }
+            .build(block_id, state.weight_threshold)
+        };
 
-        // Allow only one validation at a time
-        if state
-            .block_signatures
-            .insert(block_id.seqno, block_signatures.clone())
-            .is_err()
-        {
-            // TODO: Panic here?
-            anyhow::bail!(
-                "block validation is already in progress. \
-                session_id={}, block_id={block_id}",
-                self.inner.session_id
-            );
+        // second check for prevent data races
+        match state.block_signatures.entry(block_id.seqno) {
+            tycho_util::DashMapEntry::Occupied(_) => {
+                anyhow::bail!(
```

Comment:
Should not just remove TODO. Let's panic here: https://broxus.youtrack.cloud/articles/DNODE-A-5#task-v004